core/lib/instrument.c

```diff
@@ -6766,6 +6766,125 @@ dr_redirect_native_target(void *drcontext)
  * *loging not safe if not owning thread?
  */
 
+DR_API
+/* Schedules a shared fragment to be deleted. User must redirect control upon return
+ * using \p dr_redirect_execution() similar to what one would do when using
+ * \p dr_flush_region().
+ *
+ * As a parameter, \p tag denotes the ID of the fragment requested for deletion.
+ *
+ * Only the deletion of shared fragments is supported.
+ *
+ * No locks can be held by the caller.
+ *
+ * \return false if the function fails to schedule the deletion.
+ *
+ * \note Unlike \p dr_delete_fragment(), this function does not require the
+ * -thread_private runtime option to be set.
+ */
+bool
+dr_delete_shared_fragment(void *tag)
+{
+    dcontext_t *dcontext = get_thread_private_dcontext();
+    fragment_t *f;
+
+    if (is_couldbelinking(dcontext))
+        return false;
+
+    /* suspend threads to invalidate private ibts */
+    thread_record_t **flush_threads;
+    int flush_num_threads;
+    const thread_synch_state_t desired_state =
+        THREAD_SYNCH_SUSPENDED_VALID_MCONTEXT_OR_NO_XFER;
+    synch_with_all_threads(desired_state, &flush_threads, &flush_num_threads,
```

Comment:
Wait -- I thought the goal was to have an operation that is more lightweight than existing flushes?  A synchall is not a lightweight operation: it suspends all of the other threads.  In a many-threaded app this is a heavyweight step.  I do not think you want to do a synchall.

I thought you were going to just do an unlink and schedule for deletion.  Look at sthg like unlink_fragment_for_signal() for unlinking.

Reply:
> I thought you were going to just do an unlink and schedule for deletion. Look at sthg like unlink_fragment_for_signal() for unlinking.

You are absolutely right; that is exactly what I am trying to do. The reason I do the synchall is so I can invalidate private IBTs because (I believe) other threads could be in the fragment I am trying to delete. 

I'll have a look at unlink_fragment_for_signal

Reply:
Thanks for letting me know about unlink_fragment_for_signal. I see how unlink_fragment_for_signal is unlinking the fragment but I cannot see how it's scheduling deletion. Is the latter done automatically?

Reply:
This sounds like dr_unlink_flush_region() but for one fragment.

Reply:
Yea, pretty much.

Reply:
Ignore the unlink_fragment_for_signal: just an example of an unlink-only.  Probably this should use the dr_unlink_flush_region paths but without removing from executable areas.

Reply:
> Ignore the unlink_fragment_for_signal: just an example of an unlink-only. Probably this should use the dr_unlink_flush_region paths but without removing from executable areas.

I am having a look at this again, and now remember why I found it difficult to use the same code as dr_flush_region. Essentially, the reason boils down to the high coupling of flushing fragments and  analysis of executable areas. 

The main function I am talking about is flush_fragments_from_region(). In turn, this function calls two other sub-functions, namely flush_fragments_in_region_start() and flush_fragments_in_region_finish(), both of which use reason over executable regions and multiple fragments

What I could do is try investigate further and separate concerns, i.e, (1) the identification of fragments pertaining to an executable region and (2)  the flushing of fragments. Do you recommend me doing this please?

Another option is to try and arrange my current code, which seems to work, but I know this is not what you want and I agree in general that we should use as much shared code as possible.

It would be great if you have any advice regarding this.

Reply:
Not in a particular rush, but any thoughts/indications on the above please?

Reply:
Sorry for the delay.  I agree that the cleanest thing is to separate the flushing itself from the executable areas.  

> What I could do is try investigate further and separate concerns, i.e, (1) the identification of fragments pertaining to an executable region and (2) the flushing of fragments. Do you recommend me doing this please?

Yes, that would be my vote.